src/treelearner/monotone_constraints.hpp

```diff
@@ -78,48 +365,148 @@ class LeafConstraintsBase {
     tree_ = tree;
   }
 
- private:
+ protected:
   const Tree* tree_;
 };
 
+// used by AdvancedLeafConstraints
+struct AdvancedConstraintEntry : ConstraintEntry {
+  std::vector<AdvancedFeatureConstraints> constraints;
+
+  AdvancedConstraintEntry *clone() const final {
+    return new AdvancedConstraintEntry(*this);
+  };
+
+  void RecomputeConstraintsIfNeeded(LeafConstraintsBase *constraints_,
+                                    int feature_for_constraint, int leaf_idx,
+                                    uint32_t it_end) final {
+    if (constraints[feature_for_constraint]
+            .FeatureMinConstraintsToBeUpdated() ||
+        constraints[feature_for_constraint]
+            .FeatureMaxConstraintsToBeUpdated()) {
+      FeatureMinOrMaxConstraints &constraints_to_be_updated =
+          constraints[feature_for_constraint].FeatureMinConstraintsToBeUpdated()
+              ? constraints[feature_for_constraint].GetMinConstraints()
+              : constraints[feature_for_constraint].GetMaxConstraints();
+
+      constraints_to_be_updated.Reset(
+          constraints[feature_for_constraint].FeatureMinConstraintsToBeUpdated()
+              ? -std::numeric_limits<double>::max()
+              : std::numeric_limits<double>::max());
+
+      std::vector<int> features_of_splits_going_up_from_original_leaf =
+          std::vector<int>();
+      std::vector<uint32_t> thresholds_of_splits_going_up_from_original_leaf =
+          std::vector<uint32_t>();
+      std::vector<bool> was_original_leaf_right_child_of_split =
+          std::vector<bool>();
+      constraints_->GoUpToFindConstrainingLeaves(
+          feature_for_constraint, leaf_idx,
+          &features_of_splits_going_up_from_original_leaf,
+          &thresholds_of_splits_going_up_from_original_leaf,
+          &was_original_leaf_right_child_of_split, &constraints_to_be_updated,
+          constraints[feature_for_constraint]
+              .FeatureMinConstraintsToBeUpdated(),
+          0, it_end, it_end);
+      constraints[feature_for_constraint].ResetUpdates();
+    }
+  }
+
+  // for each feature, an array of constraints needs to be stored
+  explicit AdvancedConstraintEntry(unsigned int num_features) {
```

Comment:
I believe `num_features` could be just `int` type for the consistency with the rest codebase.